agent/grpc-external/services/resource/read.go

```diff
@@ -0,0 +1,53 @@
+package resource
+
+import (
+	"context"
+	"fmt"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+
+	"github.com/hashicorp/consul/internal/resource"
+	"github.com/hashicorp/consul/internal/storage"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+func (s *Server) Read(ctx context.Context, req *pbresource.ReadRequest) (*pbresource.ReadResponse, error) {
+	// check type exists
+	_, ok := s.registry.Resolve(req.Id.Type)
+	if !ok {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("resource type %s not registered", resource.ToGVK(req.Id.Type)))
+	}
+
+	consistency := storage.EventualConsistency
+	if isConsistentRead(ctx) {
+		consistency = storage.StrongConsistency
+	}
+
+	resource, err := s.backend.Read(ctx, consistency, req.Id)
+	if err != nil {
+		if err == storage.ErrNotFound {
+			return nil, status.Error(codes.NotFound, err.Error())
+		}
+		if _, ok := err.(storage.GroupVersionMismatchError); ok {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+		return nil, err
```

Comment:
tip: it's probably worth using [`errors.Is`](https://pkg.go.dev/errors#Is) and [`errors.As`](https://pkg.go.dev/errors#As) here in case backends [wrap the error](https://go.dev/blog/go1.13-errors).

Reply:
TIL! Fixed in https://github.com/hashicorp/consul/pull/16655/commits/d5cd617e19421a2097ed1d404cf497acbb790583